upgrade/scripts/k8s/upgrade_k8s.sh

```diff
@@ -0,0 +1,328 @@
+  scp $new_yaml_file $host:${yaml_file}
+}
+
+masters=$(grep -oP 'ncn-m\d+' /etc/hosts | sort -u)
+workers=$(grep -oP 'ncn-w\d+' /etc/hosts | sort -u)
+
+# We'll only upgrade the kubelet on the masters and the workers to the target version
+# which is the last version in the KUBERNETES_VERSIONS list
+target_version="${KUBERNETES_VERSIONS[-1]}"
+v_target_version="v${target_version}"
+minor_target_version=$(echo "$target_version" | cut -d'.' -f1-2)
+
+echo "$(prefix) Beginning Kubernetes upgrade to ${target_version}."
+
+for version in "${KUBERNETES_VERSIONS[@]}"; do
+  minor_version=$(echo "$version" | cut -d'.' -f1-2)
+
+  if [[ $(bc -l <<< "${minor_version} < 1.27") -eq 1 ]]; then
+    echo "$(prefix) ERROR Upgrade version cannot be less than v1.27. Exiting ..." >&2
+    usage >&2
+    exit 1
+  fi
+
+  # If there is one time config map work to do for a k8s version, put it here
+  if [ "$minor_version" = "1.27" ]; then
+    # Edit kubelet-config cm
+    update_kubelet_config_cm
+  fi
+
+  for host in ${masters}; do
+    # Check that current k8s version is less than or equal to the minor_version.
+    # We don't want to downgrade, so if minor_version is less than current version, do nothing.
+    current_version=$(ssh $host kubeadm version -o json | jq -r '.clientVersion.gitVersion' | cut -c 2-5)
+    if [[ $(bc -l <<< "${minor_version} > ${current_version}") -eq 1 ]]; then
+      echo "$(prefix) Installing [kubeadm-${version}] on [${host}]."
+      ssh "${host}" zypper --non-interactive install "kubeadm-${version}"
+      ssh "${host}" zypper --non-interactive install "kubernetes-cni-${k8s_cni[${minor_version}]}"
+
+      echo "$(prefix) Running kubeadm upgrade apply v${version} on [${host}]."
+      if [[ $(bc -l <<< "${minor_version} < 1.30") -eq 1 ]]; then
+        ssh "${host}" kubeadm upgrade apply -y "v${version}" --force
+      else
+        ssh "${host}" kubeadm upgrade apply -y "v${version}"
+      fi
+    else
+      echo "$(prefix) INFO Version ${minor_version} is less than ${current_version}. Don't need to upgrade kubeadm for '$host'."
+    fi
+  done
+done
+
+# Upgrade kubelet on master nodes for target_version.
+echo "$(prefix) Upgrading kubelet on master nodes to ${target_version}."
+
+for host in ${masters}; do
+  # Check current kubelet version. If less than 
+  current_version=$(ssh $host kubelet --version | awk '{ print $2 }' | cut -c 2-5)
+  if [[ $(bc -l <<< "${minor_target_version} > ${current_version}") -eq 1 ]]; then
+    if [ "$DRAIN" = "true" ]; then
+      # Drain the node, ignoring daemonsets and deleting emptydir data.
+      echo "$(prefix) Draining node: [${host}]."
+      kubectl drain "${host}" --ignore-daemonsets --delete-emptydir-data
+    fi
+
+    # Fix /etc/sysconfig/kubelet
+    fix_sysconfig "${host}"
+
+    # Update pause version in containerd
+    pause_version="pause:${pause[${minor_target_version}]}"
+    ssh "${host}" sed -i -e "s/pause:3.9/$pause_version/g" /etc/containerd/config.toml
```

Comment:
This assumes what version of pause is already there, a pattern match would make it more robust
```suggestion
    ssh "${host}" sed -i -e "s/pause:[0-9]*.[0-9]*/$pause_version/g" /etc/containerd/config.toml
```